File: C/src/z_Individuo.cpp
/***************************************************************************/
// CRISTÓBAL PÉREZ SIMÓN
// Fichero con la definición de los métodos asociadas a la clase Individuo_POO
/***************************************************************************/

#include "z_Individuo.h"
#include <cmath>
#include <iostream>

using namespace std;

/*************************************************************************/
/*************************************************************************/
// Constructor sin argumentos (crea un Individuo vacío)

Individuo_POO::Individuo_POO(){
    CAPACIDAD = vector<int>();
    NUM_ESTACIONES = 0;
    FITNESS = 0;

    KM_EXTRA = 0;
    KM_TENDENCIA = 0;
    SOBRANTE = 0;
    
    CAPACIDAD_TOTAL = 0;
    VACIAS = 0;
    LLENAS = 0;
    
    USUARIOSPERDIDOS = 0;
}

/*************************************************************************/
/*************************************************************************/
// Constructor de copia
//
// Recibe:
//      otro: Referencia a un objeto de la misma clase del que copiar 
//      los datos
//
// PRE: otro debe haber sido inicializado y ser distinto al objeto

Individuo_POO::Individuo_POO(const Individuo_POO & otro){
    if(this != &otro)
        CopiarDatos(otro);
}

/*************************************************************************/
/*************************************************************************/
// Sobrecarga del operador de asignación
//
// Recibe
//  otro: Referencia al objeto de la misma clase del que se van a copiar 
// los datos
//
//  PRE: El objeto otro debe estar inicializado y ser distinto al objeto

Individuo_POO & Individuo_POO::operator = (const Individuo_POO & otro){
    if(this != &otro){
        CopiarDatos(otro);
    }

    return (*this);
}

/*************************************************************************/
/*************************************************************************/
// Sobrecarga del operador de igualdad
// Comprueba si dos individuos son iguales (tienen el mismo fitness)
//
// Recibe:
//      otro: Referencia al objeto de la misma clase con el que comparar
//
// PRE: El objeto otro debe estar inicializado

bool Individuo_POO::operator == (const Individuo_POO & otro) const{
    return this->FITNESS == otro.FITNESS;
}

/*************************************************************************/
/*************************************************************************/
// Métodos SET para los atributos de la clase

void Individuo_POO::setCapacidad(const vector<int> & capacidad){
    CAPACIDAD = capacidad;
    NUM_ESTACIONES = capacidad.size();
    CAPACIDAD_TOTAL = sum(capacidad);
    FITNESS = 0;
    KM_EXTRA = 0;
    KM_TENDENCIA = 0;
    SOBRANTE = 0;
    VACIAS = 0;
    LLENAS = 0;
    USUARIOSPERDIDOS = 0;
}

void Individuo_POO::setNumEstaciones(const int num_estaciones){
    NUM_ESTACIONES = num_estaciones;
}

void Individuo_POO::setFitness(const float fitness){
    FITNESS = fitness;
}

void Individuo_POO::setCosteKmExtra(const float km_extra){
    KM_EXTRA = km_extra;
}

void Individuo_POO::setCosteKmsTendencia(const float km_tendencia){
    KM_TENDENCIA = km_tendencia;
}

void Individuo_POO::setSobrante(const float sobrante){
    SOBRANTE = sobrante;
}

void Individuo_POO::setCapacidadTotal(const int capacidad_total){
    CAPACIDAD_TOTAL = capacidad_total;
}

void Individuo_POO::setUsuariosPerdidos(const int usuarios_perdidos){
    USUARIOSPERDIDOS = usuarios_perdidos;
}

void Individuo_POO::setEstacionesLlenas(const int estaciones_llenas){
    LLENAS = estaciones_llenas;
}

void Individuo_POO::setEstacionesVacias(const int estaciones_vacias){
    VACIAS = estaciones_vacias;
}

/*************************************************************************/
/*************************************************************************/
// Métodos GET para los atributos de la clase

vector<int> Individuo_POO::getCapacidad() const{
    return CAPACIDAD;
}

int Individuo_POO::getNumEstaciones() const{
    return NUM_ESTACIONES;
}

int Individuo_POO::getFitness() const{
    return FITNESS;
}

int Individuo_POO::getKmExtra() const{
    return KM_EXTRA;
}

int Individuo_POO::getKmTendencia() const{
    return KM_TENDENCIA;
}

int Individuo_POO::getSobrante() const{
    return SOBRANTE;
}

int Individuo_POO::getCapacidadTotal() const{
    return CAPACIDAD_TOTAL;
}

int Individuo_POO::getUsuariosPerdidos() const{
    return USUARIOSPERDIDOS;
}

int Individuo_POO::getLlenas() const{
    return LLENAS;
}

int Individuo_POO::getVacias() const{
    return VACIAS;
}

/*************************************************************************/
/*************************************************************************/
// Método para serializar un individuo

string Individuo_POO::ToString(){
    string cadena;
    cadena += "FITNESS: " + to_string(FITNESS) + "\n";
    cadena += "KM_TENDENCIA: " + to_string(KM_TENDENCIA) + "\n";
    cadena += "KM_EXTRA: " + to_string(KM_EXTRA) + "\n";
    cadena += "SOBRANTES: " + to_string(SOBRANTE) + "\n";

    return cadena;
}

/*************************************************************************/
/*************************************************************************/
// Generacion de un nuevo individuo a partir del base, con un número de cambios
// relativo al número de estaciones y con una cantidad relativa de bicis movidas.
// Recibe:
//      tope_inferior: tope inferior de la cantidad de bicis a mover
//      config: estructura con los datos de la simulación
// Devuelve
//      Un nuevo individuo mutado

Individuo_POO Individuo_POO::mutacion(int tope_inferior, Datos_EVAL & config){

    int n_modificaciones = round(NUM_ESTACIONES * config.getPorEstaciones());
    if(n_modificaciones == 0) n_modificaciones = 1;

    Individuo_POO individuo;
    individuo.setCapacidad(CAPACIDAD);

    for(int i = 0; i < n_modificaciones; i++){
        random_device rd;
        mt19937 gen(rd());
        uniform_int_distribution<int> randint(0, NUM_ESTACIONES-1);
        int estacion = randint(gen);
        float maxcantidad = round(individuo.CAPACIDAD[estacion] * config.getPorCantidad()+1);

        uniform_int_distribution<int> randint2(-maxcantidad, maxcantidad);
        int cant = randint2(gen);
    
        if(cant != 0){
            if((individuo.CAPACIDAD_TOTAL + cant) < tope_inferior)
                cant = -cant;

            if((individuo.CAPACIDAD[estacion] + cant) < 0)
                individuo.CAPACIDAD[estacion] = 0;
            else
                individuo.CAPACIDAD[estacion] += cant;
        }
    }
    
    individuo.CAPACIDAD_TOTAL = sum(individuo.CAPACIDAD);

    return individuo;
}

/*************************************************************************/
/*************************************************************************/
// Mutación extrema de un individuo
// Recibe:
//      tope_inferior: tope inferior de la cantidad de bicis a mover
//      config: estructura con los datos de la simulación
//      n_modificaciones: número de modificaciones a realizar
// Devuelve
//      Un nuevo individuo mutado

Individuo_POO Individuo_POO::mutacion_extrema(int tope_inferior, Datos_EVAL & config, int n_modificaciones){

    Individuo_POO individuo;
    individuo.setCapacidad(CAPACIDAD);

    for(int i = 0; i < n_modificaciones; i++){
        random_device rd;
        mt19937 gen(rd());
        uniform_int_distribution<int> randint(0, NUM_ESTACIONES-1);
        int estacion = randint(gen);
        float maxcantidad = round(individuo.CAPACIDAD[estacion] * config.getPorCantidad()+1);

        uniform_int_distribution<int> randint2(-maxcantidad, maxcantidad);
        int cant = randint2(gen);

        if(cant != 0){
            if((individuo.CAPACIDAD_TOTAL + cant) < tope_inferior)
                cant = -cant;

            if((individuo.CAPACIDAD[estacion] + cant) < 0)
                individuo.CAPACIDAD[estacion] = 0;
            else
                individuo.CAPACIDAD[estacion] += cant;
        }
    }

    individuo.CAPACIDAD_TOTAL = sum(individuo.CAPACIDAD);
    return individuo;
}

/*************************************************************************/
/*************************************************************************/
// Cruce 5% de dos individuos
// Recibe:
//      uno: primer individuo con el que cruzarse
//      otro: segundo individuo con el que cruzarse
//      cap_min: capacidad mínima de las estaciones

void Individuo_POO::cruceCincoPorCiento(Individuo_POO & uno, Individuo_POO & otro, int cap_min){
    int size = min(uno.CAPACIDAD.size(), otro.CAPACIDAD.size());

    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<int> randint(1, size);
    int cxpoint1 = randint(gen);
    
    uniform_int_distribution<int>randint2(1, round((size-1)*0.05));
    int longitud = randint2(gen);

    vector<int> padre1 = uno.CAPACIDAD;
    vector<int> padre2 = otro.CAPACIDAD;
    Individuo_POO Hijo1, Hijo2;
    Hijo1.setCapacidad(uno.CAPACIDAD);
    Hijo2.setCapacidad(otro.CAPACIDAD);

    for(int x = cxpoint1; x < cxpoint1 + longitud; x++){
        int gen1 = padre1[x % size];
        int gen2 = padre2[x % size];

        if((sum(Hijo1.CAPACIDAD) + (gen2 - gen1)) < cap_min || 
            (sum(Hijo2.CAPACIDAD) + (gen1 - gen2)) < cap_min){
            break;
        }

        Hijo1.CAPACIDAD[x % size] = padre2[x % size];
        Hijo2.CAPACIDAD[x % size] = padre1[x % size];
    }

    uno = Hijo1;
    otro = Hijo2;
}

/*************************************************************************/
/*************************************************************************/
// Evaluación del individuo
// Recibe:
//      config: estructura con los datos de la simulación

void Individuo_POO::evaluacion_individuo(Datos_EVAL & config){

    config.setEvaluaciones(config.getEvaluaciones() + 1);
    LLENAS = 0;
    VACIAS = 0;

    vector<int> ocupacion(CAPACIDAD.size(), 0);

    int idestaciones = config.getDeltas()[0].size();

    /// Acumuladores de costes
    float coste_kms_extra = 0;
    float coste_kms_tendencia = 0;

    bool mov_virtual, entrada_virt;

    float aaaa;
    float movido;
    
    vector<vector<int>> DELTAS = config.getDeltas();
    vector<vector<int>> CERCA = config.getCerca();
    vector<vector<int>> TENDENCIAS = config.getTendencia();
    vector<vector<float>> CERCAKM = config.getCercaKM();

    for(int linea_delta = 0; linea_delta < DELTAS.size(); linea_delta++){
        for(int est_Origen = 0; est_Origen < idestaciones; est_Origen++){
            float pendiente = DELTAS[linea_delta][est_Origen] * config.getFactorStress();
            int nuevatendencia = TENDENCIAS[linea_delta][est_Origen];
            int it_estacion = 0;
            
            int idest = CERCA[est_Origen][it_estacion];
            
            if(pendiente == 0){
                if((ocupacion[est_Origen] == CAPACIDAD[est_Origen]) or (ocupacion[est_Origen] == 0)){
                    mov_virtual = true;
                    pendiente = nuevatendencia;
                }
                else
                    mov_virtual = false;
            }
            else
                mov_virtual = false;
            
            
            if(pendiente > (CAPACIDAD[est_Origen] - ocupacion[est_Origen]))
                USUARIOSPERDIDOS += pendiente - (CAPACIDAD[est_Origen] - ocupacion[est_Origen]);
            
            else if((ocupacion[est_Origen]+pendiente) < 0)
                USUARIOSPERDIDOS -= pendiente + ocupacion[est_Origen];
            
            while(pendiente != 0){   
                if(pendiente > 0){ // Vamos a METER est_Origen
                    if(ocupacion[idest] + pendiente <= CAPACIDAD[idest]){
                        movido = pendiente;
                        pendiente = 0;
                    }
                    else{ // NO HAY BASTANTES HUECOS
                        movido = CAPACIDAD[idest] - ocupacion[idest];
                        pendiente -= movido;
                    }

                    if(!mov_virtual){
                        ocupacion[idest] = ocupacion[idest] + movido;
                    }
                    else{
                        entrada_virt = true;
                        if(ocupacion[idest] == CAPACIDAD[idest])
                            LLENAS += 1;
                    }
                }
                else{ // Vamos a SACAR est_Origen
                    if(ocupacion[idest] + pendiente >= 0){
                        movido = -pendiente;
                        pendiente = 0;
                    }
                    else{ // FALTAN BICIS
                        movido = ocupacion[idest];
                        pendiente += movido;
                    }
                    if(!mov_virtual)
                        ocupacion[idest] -= movido;
                    else{
                        entrada_virt = false;
                        if(ocupacion[idest] == 0)
                            VACIAS += 1;
                    }
                }

                aaaa = movido * CERCAKM[est_Origen][it_estacion];
                if(mov_virtual)
                    coste_kms_tendencia += aaaa;
                else
                    coste_kms_extra += aaaa;

                
                // vamos a la siguiente
                it_estacion++;
                // si no quedan huecos en ninguna estación...
                if(it_estacion == NUM_ESTACIONES){
                    cout << "SOCOROOO... QUE NO ME CABE!!" << endl;
                    cout << "La línea delta es " << linea_delta << "-->" << DELTAS[linea_delta][0];
                    exit(-1);
                    break;
                }
                idest = CERCA[est_Origen][it_estacion];
            }
        }
    }

    // CÁLCULO DE FITNESS
    float bicis_total = (config.getTotalBicis() * (1 + config.getPorSeguridad()) * 1.0);
    SOBRANTE = (CAPACIDAD_TOTAL - bicis_total) / bicis_total;
    FITNESS = ((coste_kms_extra + coste_kms_tendencia) * config.getValorKM() + SOBRANTE * config.getFactorSobrante());
    KM_EXTRA = coste_kms_extra;
    KM_TENDENCIA = coste_kms_tendencia;
    cout << "INDIVIDUO: " << ToString(); 
}


/*************************************************************************/
/*************************************************************************/
// Copia los datos desde otro objeto de la clase
// PRE: Se ha reservado memoria para los datos

void Individuo_POO::CopiarDatos (const Individuo_POO & otro){
    CAPACIDAD = otro.CAPACIDAD;
    NUM_ESTACIONES = otro.NUM_ESTACIONES;
    FITNESS = otro.FITNESS;
    KM_EXTRA = otro.KM_EXTRA;
    KM_TENDENCIA = otro.KM_TENDENCIA;
    SOBRANTE = otro.SOBRANTE;    
    CAPACIDAD_TOTAL = sum(otro.CAPACIDAD);
    VACIAS = otro.VACIAS;
    LLENAS = otro.LLENAS;
    USUARIOSPERDIDOS = otro.USUARIOSPERDIDOS;
}

/*************************************************************************/
/*************************************************************************/
// Sumatorio de la capacidad de las estaciones

int sum(const vector<int> & capacidad){
    int sumatoria = 0;
    for(int i = 0; i < capacidad.size(); i++)
        sumatoria += capacidad[i];
    
    return sumatoria;
} 

File: Python/z_Individuo.py
#!/usr/bin/env python
# coding: utf-8

import numpy as np
import random
from math import exp

import librerias.z_io as z_io
from   librerias.z_io import debug
import sys



class Individuo_POO:
    CAPACIDAD = np.array([])
    NUM_ESTACIONES = 0
    FITNESS = 0

    KM_EXTRA = 0
    KM_TENDENCIA = 0
    SOBRANTE = 0
    TENDENCIA_ENTRADA = 0
    TENDENCIA_SALIDA = 0

    CAPACIDAD_TOTAL = 0

    # TODO GUARDAR VACIOAS LLENAS POR ESTACION
    VACIAS = 0
    LLENAS = 0

    USUARIOSPERDIDOS = 0


    def __init__(self):
        """Constructor de clase Persona"""
        self.CAPACIDAD = np.array([])
        self.NUM_ESTACIONES = 0
        self.FITNESS = 0
        self.KM_EXTRA = 0
        self.KM_TENDENCIA = 0
        self.SOBRANTE = 0
        self.VACIAS = 0
        self.LLENAS = 0
        self.CAPACIDAD_TOTAL = 0
        self.USUARIOSPERDIDOS = 0

    # def __init__(self, base):
    #     """Constructor de clase Persona"""
    #     self.CAPACIDAD = base.copy()
    #     self.NUM_ESTACIONES = len(base)
    #     self.FITNESS = 0
    #     self.KM_EXTRA = 0
    #     self.KM_TENDENCIA = 0
    #     self.SOBRANTE = 0
    #     self.VACIAS = 0
    #     self.LLENAS = 0
    #     self.CAPACIDAD_TOTAL = self.CAPACIDAD.sum()

    def setCapacidad(self,BASE):
        self.CAPACIDAD = BASE.copy()
        self.NUM_ESTACIONES = len(BASE)
        self.CAPACIDAD_TOTAL = self.CAPACIDAD.sum()
        self.FITNESS = 0
        self.KM_EXTRA = 0
        self.KM_TENDENCIA = 0
        self.SOBRANTE = 0
        self.VACIAS = 0
        self.LLENAS = 0
        self.USUARIOSPERDIDOS = 0

    def copy(self):
        copia = Individuo_POO()
        copia.CAPACIDAD       = self.CAPACIDAD.copy()
        copia.NUM_ESTACIONES  = self.NUM_ESTACIONES
        copia.FITNESS         = self.FITNESS
        copia.KM_EXTRA        = self.KM_EXTRA
        copia.KM_TENDENCIA    = self.KM_TENDENCIA
        copia.SOBRANTE        = self.SOBRANTE
        copia.VACIAS          = self.VACIAS
        copia.LLENAS          = self.LLENAS
        copia.CAPACIDAD_TOTAL = copia.CAPACIDAD.sum()
        copia.USUARIOSPERDIDOS = self.USUARIOSPERDIDOS

        return copia


    def to_dict(self):
        copia = {}
        copia["CAPACIDAD"]        = self.CAPACIDAD.copy()
        copia["NUM_ESTACIONES"]   = self.NUM_ESTACIONES
        copia["FITNESS"]          = self.FITNESS
        copia["KM_EXTRA"]         = self.KM_EXTRA
        copia["KM_TENDENCIA"]     = self.KM_TENDENCIA
        copia["SOBRANTE"]         = self.SOBRANTE
        copia["VACIAS"]           = self.VACIAS
        copia["LLENAS"]           = self.LLENAS
        copia["CAPACIDAD_TOTAL"]  = self.CAPACIDAD.sum()
        copia["USUARIOSPERDIDOS"] = self.USUARIOSPERDIDOS

        return copia





    def __str__(self):
        # Este no estaba ya cadena=str(self.CAPACIDAD)+"\n"+self.FITNESS
        #cadena = "-" * 80 + "\n"
        cadena = "FITNESS:   " + str(self.FITNESS)
        cadena += "KMS TENDENCIA: " + str(self.KM_TENDENCIA) + "\n"
        cadena += "KMS EXTRA: " + str(self.KM_EXTRA) + "\n"
        cadena += "SOBRANTES: " + str(self.SOBRANTE) + "\n"
        #cadena += "\n[" + ", ".join([str(x) for x in self.CAPACIDAD]) + "]:"
        #cadena += "CAP TOTAL: " + str(self.CAPACIDAD_TOTAL) + "\n"
        #cadena += "LLENAS:    " + str(self.LLENAS) + "\n"
        #cadena += "VACIAS:    " + str(self.VACIAS) + "\n"
        #cadena += "US. PERDIDOS: " + str(self.USUARIOSPERDIDOS) + "\n"

        #cadena += "TENDENCIA_ENTRADA: " + str(self.TENDENCIA_ENTRADA) + "\n"
        #cadena += "TENDENCIA_SALIDA:  " + str(self.TENDENCIA_SALIDA) + "\n"


        return cadena

    def mutacion(self, tope_inferior, config):
        '''
        Generacion de un nuevo individuo, a partir del base, con un numero de cambios
        relativo al numero de estaciones y con una cantidad relativa de bicis movidas.
        :param base: individuo a mutar
        :param porcent_estaciones: porcentaje de movimientos
        :param porcent_cant: porcentaje de la cantidad en función
        :return:
        '''
        minivel = 0

        debug("\n", nivel=minivel)
        debug("-." * 20, nivel=minivel)
        debug("\n", nivel=minivel)

        debug("M  U  T  A  C I O O O N N N N N N NNNNNNNN", nivel=minivel)

        n_modificaciones = round(
            self.NUM_ESTACIONES * float(config['POR_ESTACIONES']))  # 1% de las estaciones se modifican
        if n_modificaciones == 0: n_modificaciones = 1

        debug("modificaciones = ", n_modificaciones, nivel=minivel)

        individuo = Individuo_POO()
        individuo.setCapacidad(self.CAPACIDAD)

        for _ in range(n_modificaciones):
            i = random.randint(0, self.NUM_ESTACIONES - 1)
            maxcantidad = round(
                individuo.CAPACIDAD[i] * config['POR_CANT'] +1 ) # modificamos como máximo esta cantidad
            debug("MAX_CANTIDAD = ", maxcantidad, nivel=minivel)

            cant = random.randint(-maxcantidad, maxcantidad)

            debug("CANTIDAD = ", cant, nivel=minivel)

            if cant != 0:
                # Si voy a restar por debajo de la capacidad mínimo, sumar
                # if (individuo.CAPACIDAD_TOTAL + cant) < ((tope_inferior/100 + 1) * config["TOTALBICIS"]):
                if (individuo.CAPACIDAD_TOTAL + cant) < (tope_inferior):
                    cant = -cant

                if (individuo.CAPACIDAD[i] + cant < 0):
                    individuo.CAPACIDAD[i] = 0
                else:
                    individuo.CAPACIDAD[i] += cant

        individuo.CAPACIDAD_TOTAL = individuo.CAPACIDAD.sum()
        # maribel lo quita individuo.evaluacion_individuo_app_2(config)

        return individuo

    def mutacion_extrema(self, tope_inferior, config, n_modificaciones):
        '''
        Generacion de un nuevo individuo, a partir del base, con un numero de cambios
        relativo al numero de estaciones y con una cantidad relativa de bicis movidas.
        :param base: individuo a mutar
        :param porcent_estaciones: porcentaje de movimientos
        :param porcent_cant: porcentaje de la cantidad en función
        :return:
        '''
        minivel = 0

        debug("\n", nivel=minivel)
        debug("-." * 20, nivel=minivel)
        debug("\n", nivel=minivel)

        debug("M  U  T  A  C I O O O N N N N N N NNNNNNNN", nivel=minivel)

        debug("modificaciones = ", n_modificaciones, nivel=minivel)

        individuo = Individuo_POO()
        individuo.setCapacidad(self.CAPACIDAD)

        for _ in range(n_modificaciones):
            i = random.randint(0, self.NUM_ESTACIONES - 1)
            maxcantidad = round(
                individuo.CAPACIDAD[i] * config['POR_CANT'] +1 ) # modificamos como máximo esta cantidad
            debug("MAX_CANTIDAD = ", maxcantidad, nivel=minivel)

            cant = random.randint(-maxcantidad, maxcantidad)

            debug("CANTIDAD = ", cant, nivel=minivel)

            if cant != 0:
                # Si voy a restar por debajo de la capacidad mínimo, sumar
                # if (individuo.CAPACIDAD_TOTAL + cant) < ((tope_inferior/100 + 1) * config["TOTALBICIS"]):
                if (individuo.CAPACIDAD_TOTAL + cant) < (tope_inferior):
                    cant = -cant

                if (individuo.CAPACIDAD[i] + cant < 0):
                    individuo.CAPACIDAD[i] = 0
                else:
                    individuo.CAPACIDAD[i] += cant

        individuo.CAPACIDAD_TOTAL = individuo.CAPACIDAD.sum()
        # Maribel lo quita individuo.evaluacion_individuo_app_2(config)

        return individuo

    def mutacion_variable(self, tope_inferior, config, n_modificaciones):
        '''
        Generacion de un nuevo individuo, a partir del base, con un numero de cambios
        relativo al numero de estaciones y con una cantidad relativa de bicis movidas.
        :param base: individuo a mutar
        :param porcent_estaciones: porcentaje de movimientos
        :param porcent_cant: porcentaje de la cantidad en función
        :return:  cadena.append( "[" + ", ".join([str(x) for x in self.CAPACIDAD]) + "]")
        '''
        minivel = 0

        debug("\n", nivel=minivel)
        debug("-." * 20, nivel=minivel)
        debug("\n", nivel=minivel)

        debug("M  U  T  A  C I O O O N N N N N N NNNNNNNN", nivel=minivel)

        debug("modificaciones = ", n_modificaciones, nivel=minivel)

        individuo = Individuo_POO(self.CAPACIDAD)

        for _ in range(n_modificaciones):
            i = random.randint(0, self.NUM_ESTACIONES - 1)
            maxcantidad = round(
                individuo.CAPACIDAD[i] * config['POR_CANT'] +1 ) # modificamos como máximo esta cantidad
            debug("MAX_CANTIDAD = ", maxcantidad, nivel=minivel)

            cant = random.randint(-maxcantidad, maxcantidad)

            debug("CANTIDAD = ", cant, nivel=minivel)

            if cant != 0:
                # Si voy a restar por debajo de la capacidad mínimo, sumar
                # if (individuo.CAPACIDAD_TOTAL + cant) < ((tope_inferior/100 + 1) * config["TOTALBICIS"]):
                if (individuo.CAPACIDAD_TOTAL + cant) < (tope_inferior):
                    cant = -cant

                if (individuo.CAPACIDAD[i] + cant < 0):
                    individuo.CAPACIDAD[i] = 0
                else:
                    individuo.CAPACIDAD[i] += cant

        individuo.CAPACIDAD_TOTAL = individuo.CAPACIDAD.sum()
        # Maribel lo quita individuo.evaluacion_individuo_app_2(config)

        return individuo


    def mutacion_conservativa(self,config, intercambios=None, maxcantidad=5):
        #random.seed(config["SEMILLA"])

        #  numero de cambios
        longitud = len(self.CAPACIDAD)
        if intercambios == None:
            intercambios = int(longitud/4)
            # intercambios = 2

        individuo = Individuo_POO()
        individuo.setCapacidad(self.CAPACIDAD)

        for _ in range(intercambios):
            # i = valor aleatorio entre 0 y longitud
            i = random.randint(0, longitud - 1)

            # j = valor aleatorio entre 0 y longitud
            j = random.randint(0, longitud - 1)

            cant = min([random.randint(0, maxcantidad), individuo.CAPACIDAD[i], individuo.CAPACIDAD[j]])

            # Intercambio de la cantidad y los valores
            individuo.CAPACIDAD[i] += cant
            individuo.CAPACIDAD[j] -= cant
            ## -----------------------------------------------

        return individuo


    def cruce(self,otroIndividuo, cap_max, cap_min, config):
        size = min(len(self.CAPACIDAD), len(otroIndividuo.CAPACIDAD))
        cxpoint1 = random.randint(1, size)
        longitud = random.randint(1, size - 1)

        # print(cxpoint1,longitud)

        padre1 = self.CAPACIDAD.copy()
        padre2 = otroIndividuo.CAPACIDAD.copy()

        Hijo1 = Individuo_POO()
        Hijo1.setCapacidad(self.CAPACIDAD)
        Hijo2 = Individuo_POO()
        Hijo2.setCapacidad(otroIndividuo.CAPACIDAD)

        for x in range(cxpoint1, (cxpoint1+longitud)):
            gen1 = padre1[x%size]
            gen2 = padre2[x%size]


            if (Hijo1.CAPACIDAD.sum() + (gen2-gen1) ) < cap_min \
                    or \
                (Hijo2.CAPACIDAD.sum() + (gen1-gen2)) < cap_min:
                #print("CRUCE: Generación de hijo por debajo del mínimo....")
                #print(Hijo1)
                #print(Hijo2)
                break

            Hijo1.CAPACIDAD[x%size] = padre2[x%size]
            Hijo2.CAPACIDAD[x%size] = padre1[x%size]

        # print(self)
        # print(Hijo1)
        # print(Hijo2)

        return Hijo1,Hijo2
        return None



    def cruceCincoPorCiento(self, otroIndividuo, cap_min):


        size = min(len(self.CAPACIDAD), len(otroIndividuo.CAPACIDAD))
        cxpoint1 = random.randint(1, size)
        longitud = random.randint(1, round((size - 1)*0.05))
        #print("Longitud del cruce:", longitud)

        # print(cxpoint1,longitud)

        padre1 = self.CAPACIDAD.copy()
        padre2 = otroIndividuo.CAPACIDAD.copy()

        Hijo1 = Individuo_POO()
        Hijo1.setCapacidad(self.CAPACIDAD)
        Hijo2 = Individuo_POO()
        Hijo2.setCapacidad(otroIndividuo.CAPACIDAD)

        for x in range(cxpoint1, (cxpoint1 + longitud)):
            gen1 = padre1[x % size]
            gen2 = padre2[x % size]

            if (Hijo1.CAPACIDAD.sum() + (gen2 - gen1)) < cap_min \
                or \
                (Hijo2.CAPACIDAD.sum() + (gen1 - gen2)) < cap_min:
                # print("CRUCE: Generación de hijo por debajo del mínimo....")
                # print(Hijo1)
                # print(Hijo2)
                break

            Hijo1.CAPACIDAD[x % size] = padre2[x % size]
            Hijo2.CAPACIDAD[x % size] = padre1[x % size]

        return Hijo1, Hijo2


    def evaluacion_individuo(self, config):
        '''
        Evaluación del individuo USANDO UNA APP móvil que le dice cuál es la estación libre más cercana.
        :param config:
        :return:
        '''
        cont_while=0;
        minivel = 0
        config["EVALUACIONES"] = config["EVALUACIONES"]  + 1
        i = 0
        self.LLENAS = 0
        self.VACIAS = 0
        ocupacion = np.zeros_like(self.CAPACIDAD)
        IDESTACIONES = range(len(config["DELTAS"][0]))#range(self.NUM_ESTACIONES)

        ### Acumuladores de costes
        coste_kms_extra = 0
        coste_kms_tendencia = 0

        self.TENDENCIA_ENTRADA = 0
        self.TENDENCIA_SALIDA = 0

        for linea_delta in range(len(config["DELTAS"])):
            #print("\nLinea_delta:",linea_delta, "-> deltas[75:85]:",config["DELTAS"][linea_delta])
            for est_Origen in IDESTACIONES:
                pendiente = config["DELTAS"][linea_delta, est_Origen] * config[   "FACTOR_STRESS"]  # numero de est_Origen a depositar
                nuevatendencia = config["TENDENCIA"][linea_delta, est_Origen]
                it_stacion = 0  # numero de orden (n) de la mas cercana
                idest = config["CERCA"][est_Origen, it_stacion]  # nombre de la n estacion mas cercana
                if pendiente == 0:
                    if ocupacion[est_Origen] == self.CAPACIDAD[est_Origen] or ocupacion[est_Origen] == 0:
                        mov_virtual = True
                        pendiente = nuevatendencia
                    else:
                        mov_virtual= False
                else:
                    mov_virtual = False

                if pendiente > (self.CAPACIDAD[est_Origen] - ocupacion[est_Origen]):
                    self.USUARIOSPERDIDOS += pendiente - (self.CAPACIDAD[est_Origen] - ocupacion[est_Origen])
                elif (ocupacion[est_Origen]+pendiente) < 0:
                    self.USUARIOSPERDIDOS -= pendiente + ocupacion[est_Origen] #### HHHH AQUI SERÍA += EN VEZ DE -=
                while pendiente != 0:
                    if pendiente > 0:  ### Vamos a METER est_Origen
                        if ocupacion[idest] + pendiente <= self.CAPACIDAD[idest]:  ## HAY HUECOS SUFICIENTES
                            movido = pendiente
                            pendiente = 0
                        else:  ## NO HAY HUECOS BASTANTES
                            #if(est_Origen==80 and linea_delta==172):
                            #        print("aquí empieza a generar:", pendiente)
                            movido = self.CAPACIDAD[idest] - ocupacion[idest]
                            pendiente = pendiente - movido

                        if not mov_virtual:
                            ocupacion[idest] = ocupacion[idest] + movido
                        else:
                            entrada_virt = True
                            if ocupacion[idest] == self.CAPACIDAD[idest]:
                                self.LLENAS += 1

                    else:  ### Vamos a SACAR est_Origen
                        if ocupacion[idest] + pendiente >= 0:  ## HAY SUFICIENTES
                            movido = -pendiente
                            pendiente = 0
                            #print("\t Ocupación[idest]=",ocupacion[idest])

                        else:  ## FALTAN BICIS
                            movido = ocupacion[idest]
                            pendiente = pendiente + movido

                        if not mov_virtual:
                            ocupacion[idest] = ocupacion[idest] - movido
                        else:
                            entrada_virt = False
                            if ocupacion[idest] == 0:
                                self.VACIAS += 1
                                # tendencia = abs(config["TENDENCIA"][linea_delta, est_Origen])  # TODO cambiar

                    aaaa = movido * config["CERCA_KM"][est_Origen, it_stacion]
                    if mov_virtual:
                        coste_kms_tendencia = coste_kms_tendencia + aaaa
                        if entrada_virt:
                            self.TENDENCIA_ENTRADA = self.TENDENCIA_ENTRADA + aaaa
                        else:
                            self.TENDENCIA_SALIDA = self.TENDENCIA_SALIDA + aaaa
                    else:
                        coste_kms_extra = coste_kms_extra + aaaa
                        #if aaaa > 0 :
                        #    print("\nKM: cercanas_kms[",est_Origen,",",it_stacion,"]:",config["CERCA_KM"][est_Origen, it_stacion], " movido:",movido, " se suman,",aaaa," a coste_kms_extra:",coste_kms_extra-aaaa)

                    # vamos a la siguiente
                    it_stacion += 1
                    # si no quedan huecos en ninguna estacion...
                    if it_stacion == self.NUM_ESTACIONES:
                        print("SOCORROOO... QUE NO ME CABE!!")
                        print("La línea delta es ", linea_delta, "-->", config["DELTAS"][linea_delta])
                        sys.exit(-1)
                        break
                    idest = config["CERCA"][est_Origen, it_stacion]
                    #print("Estación de destino:",idest, " consultada desde est_Origen:",est_Origen, " it_station:",it_stacion)

            debug("RESULTADO:        ", "\t".join([str(x) for x in ((self.CAPACIDAD - ocupacion))])[:], nivel=minivel)

        ## CALCULO DE FITNESS
        #bicis_total = np.round(config["TOTALBICIS"] * (1 + config["POR_SEGURIDAD"]) * 1.0)
        bicis_total = (config["TOTALBICIS"] * (1 + config["POR_SEGURIDAD"]) * 1.0)
        plazas_sobrantes = (self.CAPACIDAD_TOTAL - bicis_total) / bicis_total
        self.SOBRANTE = plazas_sobrantes  ##
        fitness = ((coste_kms_extra + coste_kms_tendencia) * config["VALOR_KM"] + self.SOBRANTE * config["FACTOR_SOBRANTE"])
        self.FITNESS = fitness
        self.KM_EXTRA = coste_kms_extra
        self.KM_TENDENCIA = coste_kms_tendencia
        print("INDIVIDUO:", self)
        # sys.exit(-1)

    def fila(self):
        cadena = []
        cadena.append( "[" + ", ".join([str(x) for x in self.CAPACIDAD]) + "]")
        cadena.append(str(self.CAPACIDAD_TOTAL))
        cadena.append(str(self.FITNESS))
        cadena.append(str(self.KM_EXTRA))
        cadena.append(str(self.KM_TENDENCIA))
        cadena.append(str(self.SOBRANTE))
        cadena.append(str(self.LLENAS))
        cadena.append(str(self.VACIAS))
        cadena.append(str(self.USUARIOSPERDIDOS))
        return cadena


